internal/uvm/mem_allocator.go

```diff
@@ -0,0 +1,182 @@
+func getMemoryClassType(s uint64) uint32 {
+	s = s >> 20
+	memCls := uint32(0)
+	for s > 1 {
+		s = s >> 2
+		memCls++
+	}
+	return memCls
+}
+
+// getMemoryClassSize returns size in bytes for a given memory class
+func getMemoryClassSize(memCls uint32) uint64 {
+	switch memCls {
+	case MemCls1MB:
+		return 1 * MegaByte
+	case MemCls4MB:
+		return 4 * MegaByte
+	case MemCls16MB:
+		return 16 * MegaByte
+	case MemCls64MB:
+		return 64 * MegaByte
+	case MemCls256MB:
+		return 256 * MegaByte
+	case MemCls1GB:
+		return 1024 * MegaByte
+	case MemCls4GB:
+		return 4 * 1024 * MegaByte
+	}
+	return 0
+}
+
+type memSlot struct {
+	memCls uint32
+	offset uint64
+	size   uint64
+	next   *memSlot
+}
+
+type memoryAllocator struct {
+	slots [MemClsNum]*memSlot
+	busy  map[uint64]*memSlot
+}
+
+// newDefaultMemoryAllocator returns an allocator with a single 0-offset 4GB memory slot
+func newDefaultMemoryAllocator() memoryAllocator {
+	mem := memoryAllocator{
+		busy: make(map[uint64]*memSlot),
+	}
+	mem.slots[MemClsNum-1] = &memSlot{
+		offset: 0,
+		size:   4 * 1024 * MegaByte,
+	}
+	return mem
+}
+
+func (mc *memoryAllocator) findFreeOffset(memCls uint32) (uint64, error) {
+	memClsCount := uint32(len(mc.slots))
+	if memCls >= MemClsNum {
+		return 0, ErrInvalidMemoryClass
+	}
+
+	for i := memCls; i < memClsCount; i++ {
+		if ms := mc.slots[i]; ms != nil {
+			return ms.offset, nil
+		}
+	}
+	return 0, ErrNotEnoughSpace
+}
+
+// allocate returns first available memory slot for a given `memCls` size class. If none left,
+// expands the existing memCls
+func (mc *memoryAllocator) allocate(memCls uint32) (*memSlot, error) {
+	if memCls >= MemClsNum {
+		return nil, ErrInvalidMemoryClass
+	}
+
+	alloc := func() *memSlot {
+		s := mc.slots[memCls]
+		if s != nil {
+			mc.slots[memCls] = s.next
+			s.next = nil
+			mc.busy[s.offset] = s
+			return s
+		}
+		return nil
+	}
+
+	if slot := alloc(); slot != nil {
+		return slot, nil
+	}
+
+	if err := mc.expand(memCls); err != nil {
+		return nil, err
+	}
+
+	slot := alloc()
+	if slot != nil {
+		return slot, nil
+	}
+
+	return nil, ErrNotEnoughSpace
+}
+
+// release returns resources back, also making sure that the `offset` and `memCls` are valid
+func (mc *memoryAllocator) release(memCls uint32, offset uint64) error {
+	ms, ok := mc.busy[offset]
+	if !ok {
+		return errors.Errorf("no memory allocated at offset: %d", offset)
+	}
+
+	if ms == nil {
+		return errors.Errorf("expected memory slot at offset: %d, found nil instead", offset)
+	}
+
+	if memCls != ms.memCls {
+		return errors.Errorf("invalid release request. memory slot types mismatch: actual=%d, requested=%d", ms.memCls, memCls)
+	}
+
+	delete(mc.busy, offset)
+	ms.next = mc.slots[ms.memCls]
+	mc.slots[ms.memCls] = ms
+	return nil
+}
+
+// expand breaks down the next memory class into smaller memCls type segments
+func (mc *memoryAllocator) expand(memCls uint32) error {
```

Comment:
We have this function for expanding memory class slots, but we never try to coalesce slots into a larger memory class size, do we? I feel like this may be overkill for the number of vpmem devices we plan on having, but there could be a scenario where we've broken down all the larger memory class size slots (using expand) and no longer will have any place to put a request of a large amount of memory even though the space exists.

Reply:
with the current simplified approach it's not super trivial, but I can probably work on the actual buddy implementation, which will probably make the merging easier

Reply:
what do you mean by buddy implementation?

Reply:
What do you anticipate the changes to be like? I was thinking just a doubly linked list and then checking for if previous or next have contiguous memory to the current one on release

Reply:
maybe I was overthinking this, but the way I saw it is we want to merge things that were split in the first place, so we'd need to keep track of the "parent-child" relationship. using just a contiguous memory might not work in all cases, since we could end up merging contiguous region that overlaps 2 parents.
so instead, I was going to have a `free` and `busy` "pool" for each class and update memory region struct to also have a pointer to parent. in that case I won't use linked lists. this is less memory efficient, but the largest map can be 4096 elements anyway, which I don't think is too much.